### ISSUE TITLE
[RFC] Hyper Parameters Tuning for Quantized Index with MemoryOptimization Enabled

### ISSUE THREAD

**Issue by 0ctopus13prime** [opensearch-project/k-NN#2877]
# Hyper Parameters Tuning for Early Termination

# 1. Goal

This document aims to provide a clear problem definition of recall drop with memory optimized search enabled. It also seeks to clearly state the goal of the tuning problem along with the corresponding objective function.

# 2. Recall Drops With Memory Optimize Search

## 2.1. Optimistic search recap
Optimistic search consists of two phases. In the first phase, it runs an approximate HNSW graph search as usual, but uses a proportional k value for each segment based on its size. In the second phase, it performs a deep-dive search.

The theory behind the first-phase search is based on binomial probability. Each segment’s contribution to the global top-k results is assumed to be proportional to its graph size. The key question is: how large should k be for each segment to ensure that all of its potential contributions to the global top-k are captured?

For example, suppose one segment contains 100,000 documents, the total number of documents is 1,000,000, and the global k is 50. The segment’s expected contribution probability is p = 100,000 / 1,000,000 = 1 / 10. Under a binomial model, the expected number of documents this segment contributes to the final top-k is E(X) = k * p = 5.

However, this is only the expectation. In rare cases, the segment may contribute significantly more than 5 documents, potentially even 20. To account for such tail cases, we must request more than the expected value.

Using the binomial distribution, a 95% confidence bound can be approximated as:

k * p + 1.645 · sqrt(k * p * (1 − p))

With k = 50 and p = 1 / 10, this yields approximately 9 documents. In theory, collecting the top-9 results from this segment should include its contribution to the global top-k with 95% confidence.

In practice, however, experiments by Lucene committers show that this theoretical bound is insufficient. Empirically, valid results are observed only when the multiplier is around 16 rather than 1.645. For this reason, we use a factor of 16 instead of the standard normal approximation coefficient.

```
private static int perLeafTopKCalculation(int k, float leafProportion) {
    // E(X) = k * leafProportion
    // Std = Math.sqrt(k * p * q) = Math.sqrt(k * leafProportion * (1 - leafProportion))
    // With above example, it will return 38.
    return (int) Math.max(1, k * leafProportion + LAMBDA * Math.sqrt(k * leafProportion * (1 - leafProportion)));
}
```

In the second phase, a deep-dive search is triggered only for promising segments whose minimum score exceeds the minimum k-th score among the results collected in the first phase (that is, the score of the last document in the current top-k). Note that this phase runs graph search starting from the results collected in the first phase, operating on the bottom graph layer (that is, the dense graph). This approach allows the search to retrieve additional competitive documents from selected segments, resulting in higher overall recall.

Please see
- https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/query/nativelib/NativeEngineKnnVectorQuery.java#L271
- https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/lucene/ReentrantKnnCollectorManager.java
- https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/lucene/OptimisticKnnCollectorManager.java

## 2.2 2-Phased search overview
For a quantized index, the search first performs an HNSW graph search with an expanded k, followed by an exact search using full-precision vectors to refine the results. For example, consider k = 100, vector dimension = 768, and three segments. In the first phase, the search uses k = 200 per segment and collects the results. After this phase, there are 600 candidates in total (200 × 3). The search then performs an exact search using full-precision vectors over these 600 candidates and selects the final top-100 results.

# 3. Optimization Problem

This is a multi-objective optimization problem. The idea is to increase the algorithm’s greediness (i.e., aggressiveness), allowing it to visit more nodes in order to make up for the dropped recall. Consequently, some additional latency is inevitable. The objective, therefore, is to find an optimal `kExpansion` that restores recall to the same level as Faiss C++ while incurring the minimum possible cost — for example, execution time.

## 3.1. Hyper Parameters
* kExpansion: New `k` for internal vector search would be Faiss efSearch * `kExpansion`. 
For each rescoring enabled request, [RescoreContext](https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/query/rescore/RescoreContext.java) will be created. 
And at the first phase search, the user provided `k` is expanded to `k*` which is greater than `k` in [RescoreContext#getFirstPassK](https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/query/rescore/RescoreContext.java).

To match up the recall drop when memory optimized search is enabled, we need further expand the already expanded `k`.
This optimization problem is to determine how much it should be bigger to match up the recall while minimizing search cost impact.

## 3.2. Objective Function

We need to minimize both recall errors and the number of nodes visited.

1. Minimize |Faiss C++ recall - LuceneOnFaiss recall|
2. Minimize search cost
  2.1. Minimize Node Visits
  2.2. Execution time

We can simply start with a simple penalty model
```
def objective(baseline_recall: float,
              baseline_execution_time: float,
              new_recall: float,
              new_execution_time: float) -> float:
    # Recall regression penalty
    recall_penalty = max(0.0, baseline_recall - new_recall)

    # Relative execution time regression penalty
    time_penalty = max(0.0, (new_execution_time / baseline_execution_time) - 1.0)

    # Squared sum of penalties
    loss = recall_penalty**2 + time_penalty**2
    return loss
```

When aggregating losses, we can simply pick the maximum one.

# 4. Implementation Notes
I’m considering using [Optuna](https://optuna.org/) to tune the hyperparameters. It’s straightforward to implement, which allows us to focus on applying the two parameters in OpenSearch. Since Optuna leverages Bayesian optimization, it generally finds optimal configurations more efficiently than grid search or similar methods.

## 4.0. Experiment Set-up Overview
<img width="1276" height="560" alt="Image" src="https://github.com/user-attachments/assets/67d095b1-1cc5-465a-89c6-df87b67ffbd3" />

One suite consists of an OSB Node and an OpenSearch Node. The OSB Node runs a Flask API server, which handles communication with the OptimizerCoordinator. When a request is received, the OSB Node runs the benchmark, collects metrics, and returns them to the coordinator. Each suite has an index built with a specific configuration (for example: Cohere-10M, 8×, 6 shards, force-merged). There are two identical suites: one represents the baseline using Faiss C++, and the other represents the new system. The built index is shared between the two suites.

## 4.1. OpenSearch Set up
Optuna provides the next values for `expansion ratio`. Therefore, we need a way to apply  the new value at runtime. To achieve this in [here](https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/query/rescore/RescoreContext.java#L81) that loads the values from a JSON file.

```
// JSON Example

{
  "kExpansion": ...  // Ex: 1.5 indicates new k would be k * 1.5.
}

```

```
// New class for parameters
@Data
public class ExperimentParameters {
    private static final String CONFIG_PATH = "/efs/exp/parameters.json";
    public static ExperimentParameters PARAMS = null;

    public float kExpansion;

    public static void load() {
        ObjectMapper mapper = new ObjectMapper();
        byte[] jsonBytes = Files.readAllBytes(Path.of(CONFIG_PATH));
        PARAMS = mapper.readValue(jsonBytes, ExperimentParameters.class);
    }
}

// In RescoreContext
    public int getFirstPassK(int finalK, boolean isShardLevelRescoringDisabled, int dimension) {
        // Only apply default dimension-based oversampling logic when:
        // 1. Shard-level rescoring is disabled
        // 2. The oversample factor was not provided by the user
        if (isShardLevelRescoringDisabled && !userProvided) {
            // Apply new dimension-based oversampling logic when shard-level rescoring is disabled
            if (dimension >= DIMENSION_THRESHOLD_1000) {
                oversampleFactor = OVERSAMPLE_FACTOR_1000;  // No oversampling for dimensions >= 1000
            } else if (dimension >= DIMENSION_THRESHOLD_768) {
                oversampleFactor = OVERSAMPLE_FACTOR_768;   // 2x oversampling for dimensions >= 768 and < 1000
            } else {
                oversampleFactor = OVERSAMPLE_FACTOR_BELOW_768;  // 3x oversampling for dimensions < 768
            }
        }

        // Expand `k` further for memory optimized search.
        oversampleFactor = oversampleFactor * ExperimentParameters.PARAMS.kExpansion;

        // The calculation for firstPassK remains the same, applying the oversample factor
        return Math.min(MAX_FIRST_PASS_RESULTS, Math.max(MIN_FIRST_PASS_RESULTS, (int) Math.ceil(finalK * oversampleFactor)));
    }
```

### [Optional] 4.1.1. Tracking Number of Nodes Visited

To minimize computation costs when compensating for dropped recall, Optuna must be able to retrieve the total number of nodes visited during each evaluation round. For this, we can maintain a static atomic integer to track the number of visited nodes. Additionally, we need to ensure that only unique node IDs are counted. This requires a small tweak in FaissHnswGraph. Lucene’s HNSW logic first traverses from the top down to layer 1, then resets its bitset before continuing greedy BFS in the bottom layer 0. Therefore, to ensure correctness, we need to track unique node IDs separately per layer.

```
public class FaissHNSWGraph extend HnswGraph {
    public static AtomicLong NUM_VISITED_NODES = new AtomicLong(0);

    ...

    @Override
    public int nextNeighbor() {
        if (nextNeighborIndex < numNeighbors) {
            // vvvvvvvvvvvvvv Increase node visit
            NUM_VISITED_NODES.incrementAndGet();
            return neighborIdList[nextNeighborIndex++];
        }

        // Neighbor list has been exhausted.
        return NO_MORE_DOCS;
    }
}
```

### 4.1.2. Stats API modification
We also need a small tweak in the KNN stats API to include the number of nodes visited.

```

// GET /_plugins/_knn/stats
public class KNNStatsTransportAction extends TransportNodesAction {
    ...

    @Override
    protected KNNStatsResponse newResponse(
        KNNStatsRequest request,
        List<KNNStatsNodeResponse> responses,
        List<FailedNodeException> failures
    ) {

        Map<String, Object> clusterStats = new HashMap<>();
        Set<String> statsToBeRetrieved = request.getStatsToBeRetrieved();

        for (String statName : knnStats.getClusterStats().keySet()) {
            if (statsToBeRetrieved.contains(statName)) {
                clusterStats.put(statName, knnStats.getStats().get(statName).getValue());
            }
        }
        
        // Add #nodes visit
        Map<String, Object> newClusterStats = new HashMap<>(clusterStats);
        newClusterStats.put("$OS_NUM_NODES_VISIT", FaissHNSWGraph.NUM_VISITED_NODES.get());
        // Reset the variable
        FaissHNSWGraph.NUM_VISITED_NODES.set(0);
        // Load parameters
        ExperimentParameters.load();

        return new KNNStatsResponse(clusterService.getClusterName(), 
                                    responses, 
                                    failures, 
                                    newClusterStats);
    }
}

```

## 4.2. Update Parameters in Optuna

We can mount an NFS volume to each OpenSearch container. Before sending queries to evaluate recall, a JSON file in the mounted directory can be updated. On the next request cycle, OpenSearch can then load the new parameters for serving queries.

```
import optuna
import random
import json

# Repeat 1,000 times
N = 1000

# Answer set to match
BASELINE_METRICS = None

# OpenSearch endpoints. Each endpoint has different model
OPENSEARCH_ENDPOINTS = [...]

def run_osb_measure(opensearch_endpoints, faiss_recalls):
    metrics = []
    for idx, endpoint in enumerate(opensearch_endpoints):
        # Trigger traffic generation
        run_osb_async(endpoint)

    for idx, endpoint in enumerate(opensearch_endpoints):
        ... collect metrics and append it to metrics

        recalls.append(acquiredRecall@k)
        recall_diff.append(abs(acquiredRecall@k - faiss_recalls[idx]))
    return metrics

def do_evaluate(baseline_metric, acquired_metric):
    ... # measure errors on recall + search cost

# Example "simulation": given two hyperparameters,
# produce an "output" and a "cost".
def evaluate(k_expansion, greediness):
    # Update parameters first
    parameters = {
        'kExpansion': k_expansion
    }
    with open('/efs/exp/parameters.json', 'wt') as f:
        f.write(json.dumps(parameters))

    # Run one batch search
    metrics = run_osb_measure(opensearch_endpoints, faiss_recalls)
    
    # Evaluate the results with the baseline
    errors = 0
    for i in range(len(BASELINE_METRICS)):
        errors += do_evaluate(BASELINE_METRICS[i], metrics[i])
    return errors


# Multi-objective objective function
def objective(trial):
    k_expansion = trial.suggest_float("kExpansion", 1, 5)

    # We have to minimize error and costs
    return evaluate(k_expansion, greediness)

# We first need to collect baseline metrics
# This recall will be acquired via OSB with memory optimized search disabled.
# e.g. Faiss C++ will give recall.
BASELINE_METRICS = ...

# Create a multi-objective study. Our goal is to minimize errors + costs.
study = optuna.create_study(directions=["minimize"])
study.optimize(objective, n_trials=N)

# Get Pareto-optimal trials
print("Number of Pareto-optimal solutions:", len(study.best_trials))
for t in study.best_trials:
    print(f"kExpansion={t.params['kExpansion']}, error={t.values[0]:.2f}, cost={t.values[1]:.2f}")
```

### 4.3. Flask Server in OSB Node
Once a request is received, it will do

1. Reset and make engine load new parameter via GET /_plugins/_knn/stats
2. For memory optimized case, it should run warm-up.
3. Run OSB
4. Collect metrics
5. Return metrics back to coordinator

### 4.4. Script
1. Fork new environment (OSB Node + Engine)
  - Baseline (Using Faiss C++)
  - Candidate (MemoryOptimizedSearch is enabled)
2. Trigger indexing in Baseline
3. Copy the built index from Baseline to Candidate
4. Create optimization coordinator

# 5. Tuning Plans
Since this is the search parameters tuning plan, we can simply build all indices required prior to this experiment set-up then proceed experiment. 

## 5.1. Dimensions
- [P0] Vector Dimension : Since we're giving different oversample factor values depending on dimension value, we need to ensure that all are covered. (At the moment, we have three cases: D < 768, 768 <= D < 1000, D >= 1000)
- [P0] Single segment VS multi segment
- [P0] Compression ratio : 
  - 8x
  - 16x
  - 32x
- [P1] ADC
- [P1] Nested case
- [P1] Filtering ratio : 0.1%, 1%, 5%, 10%, 25%, 50%, 75%, 99%

## 5.2. Plan
1. Experiment environment set-up
  1.1. Build all indices required.
  1.2. Build Optuna component.
  1.3. Configure OSB Node 
2. Experimenting.
  2.1. Start with Cohere-1M, Cohere-10M
  2.2. Include more data set (note that we need dataset whose dimension is (D < 768, 768 <= D < 1000, 1000 <= D)
  2.3. Incorporate more cases (ex: nested, filtering)
  2.4. Expand the experiment over other data-set and dimensions.

**Comment by 0ctopus13prime** [comment]
Hi @heemin32 

Q : If I understand correctly, there are two queues: a global queue and a local queue. At the moment, when the user provides k, both queues are assigned a size of k. We would now like to adjust the sizes of these two queues independently, using kExpansion to update the global queue size and greediness to update the local queue size?

Yes! That's the idea. I found that increasing the value from k=100 to k=150, LuceneOnFaiss not matches up with Faiss C++ recall. But also latency came down along with it. So I'm thinking maybe we could adjust two parameters to see if we can still match up the recall with the minimum latency degradation.

Eventually, tuned parameters will be used for 32x as it's showing -5% degradation with 2x throughput improvement, and -5% recall drop seems huge.

**Comment by andrross** [comment]
[Catch All Triage](https://github.com/opensearch-project/.github/blob/main/TRIAGING.md) - [1](https://github.com/krisfreedain) [2](https://github.com/andrross) [3](https://github.com/cwperks) [4](https://github.com/swatikothari)

**Comment by 0ctopus13prime** [comment]
Hey @MrFlap 
I think we need to review whether the recall can be improved with lesser efforts when optimistic search comes in place - https://github.com/opensearch-project/k-NN/issues/2924
Optimistic search also has a parameter 'lambda', if it turns out optimistic can perform better, than we might consider tuning 'lambda' instead of `k` and 'greediness'
